### PR TITLE
Add support for sentinel and length refinement of tuples

### LINKED ISSUE #4050: Bug: unable to refine tuples of different arity
I would expect the following to work:

```ts
// @flow

type ImportantThing = string;
type SomeTuple = [ImportantThing];
type OtherTuple = [*, ImportantThing, *];

function getImportantThing (args: SomeTuple | OtherTuple) {
	return args[1] ? args[1] : args[0];
}
```

but instead I get:

```
8: 	return args[1] ? args[1] : args[0];
           ^ access of computed property/element. Out of bound access. This tuple has 1 elements and you tried to access index 1 of
8: 	return args[1] ? args[1] : args[0];
           ^ tuple type
```

I've been unable to find a way to refine tuples of different arity. (Hope I'm not missing something obvious.)

### LINKED ISSUE #4377: Type refinements for tuples like with objects
It'd be great if this worked [(try it!)](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgByYAvGANpgBEAdlQDRg0CuAtgEZ4BOYAuprgIBlUhWoBnBmHEYuASxoBzPuijMaAYwxy4NMFAAUNcWABcRMAB8wQgJRmwANzhyAJmADeqMGDlQwRuLkAAy8pCRktFT2Xj4+ujBYhGycXOKB5ACMvLbeYAC+YHgw4gR+AcYhYRGRkjF5AJAJWEKyCorpldm5Pvmofahqmtq6YM3JHNzpAB4OLJNcMX1DWjp6za3ySukm5jJbiktAA):

```
/* @flow */

type N = [ "n", number ]
type S = [ "s", string ]

function f(ns : N | S) : void {
  if (ns[0] === "n") {
    onlyNumbers(ns[1])
  } else if (ns[0] === "s") {
    onlyStrings(ns[1])
  }
}

function onlyNumbers(x : number) {}
function onlyStrings(s : string) {}
```

But it errors:

```
8:     onlyNumbers(ns[1])
                   ^ string. This type is incompatible with the expected param type of
14: function onlyNumbers(x : number) {}
                             ^ number
10:   	onlyStrings(ns[1])
                   ^ number. This type is incompatible with the expected param type of
15: function onlyStrings(s : string) {}
                             ^ string
```

### LINKED ISSUE #4556: Bug: Refinement of tuple length does not work
[Try flow](https://flow.org/try/#0PTAEAEDMBsHsHcBQAXAngBwKagPIDtsBeUAbQEM9UBdFDbAFXllGPMoBpQLrFFIBXPAGNkAS1h5QEzDgBOjWAApk-dNEwAuXAVAAfUAoCUoAN6JQoUZFDLV6gHTq8Ac2QALFoWIAmY2YsWKmqYJACMNBYAvoiRQA)

```js
// @flow
type One = [any]
type Two = [any, any]

function oneOrTwo(tuple: One | Two) {
  if (tuple.length === 2) {
    tuple[1] // results in flow error, but shouldn't
  }
}
```

### LINKED ISSUE #4873: Flow doesn't refine tuple type by element
Simplified example:

```javascript
type MyExpressionMeta = {
    name: string;
};
type MyExpressionNode = ['hbs:expression', MyExpressionMeta, []];

type MyDOMValue = string | MyExpressionNode;
type MyDOMKey = string | MyExpressionNode;
type MyDOMNodeAttributes = Array<[MyDOMKey, MyDOMValue]>;
type MyDOMNode = ['#text', MyDOMNodeAttributes, Array<MyNode>];
type MyNode = MyDOMNode | MyExpressionNode
                  
function generate(node: MyNode): string {
    if (node[0] === 'hbs:expression') {
        return node[1].name;
    }

    return '';
}
```

[Reproducible case in Try Flow](https://flow.org/try/#0C4TwDgpgBAsiCiAPMAnCBndBLA9gOxgmAEMoBeKAbwCgoo9iBbCALinWBSzwHMBuagF8BoSLATI0mXHgByOACbQKAbQDkACwBG6FhEkZs+NQBpxSVIZmESZlQF17A6qOhwAIgHkYANWIAbAFdldk5uHigAH3MDaXx5JRFwNxAvGABpCBByUK5eKJjLOLlFCCSxD28EiABBYDCtQOAMHJqUFGIQAB4VSoyssz6-IIh7AD5ylLTqnPUAYmbEYFNxadK6hqaMMzaO7rhqsacXZPEZij6Z6LgLKSMSpVo6Z5fX5+oAM0C8AGNgGSgPAgeAgHWaAAo8KU2AdSgBKNgcPIRGgASCwHygkNKKgADPZyGQKJodHpYvc1HCqE83mhgIEUHh6DiAIz2AB0DGYAhegmo1FRdIZTLUagEgiAA)

What I'm expecting is that Flow recognizes that the first element can be one of two tuples, and that checking the first value of the tuple refines the expected type to only be `MyExpressionNode`, whose second value is `MyExpressionMeta` which has the `name` property I'm trying to read.

### PR DESCRIPTION
**Pull request by popham**
Fixes #4050, fixes #4377, fixes #4556, and fixes #4873.

### CODE REVIEW

**Comment by bradennapier** [comment]
+1 !  Definitely need this!! :-)

**Comment by anru** [comment]
it's absolutely killer feature for flow.

@vkurchatkin @mroch could we get merged this PR into the codebase ?

**Comment by PeteJodo** [comment]
I'd like to use this for better refining React children. To give a semi-contrived example:
```jsx
type LayoutOneProps = {
    layoutOne: boolean,
    children: [
        React.Element<typeof Level>,
        React.Element<typeof Sidebar>,
        React.Element<typeof Content>
    ],
};

type LayoutTwoProps = {
    layoutTwo: boolean,
    // currently required to have `null` as the last value -------------------------v
    children: [React.Element<typeof LeftColumn>, React.Element<typeof RightColumn>, null],
};

const LayoutWrapper = (props: LayoutOneProps | LayoutTwoProps) => {
    if (props.layoutOne) {
        return (
            <div className="container">
                <div className="row">
                    {props.children[0]}
                </div>
                <div className="row">
                    {props.children[1]}
                    {props.children[2]}
                </div>
            </div>
        );
    } else if (props.layoutTwo) {
        return (
            <div className="container">
                <div className="row">
                    {props.children}
                </div>
            </div>
        );
    }
};

<LayoutWrapper layoutTwo>
    <LeftColumn />
    <RightColumn />
    {null /* currently required to do this */}
</LayoutWrapper>
```

File: src/typing/refinement.ml [hunk 1]
```diff
@@ -29,18 +29,33 @@ let rec key = Ast.Expression.(function
...
-  (* foo.bar.baz -> Chain [Id baz; Id bar; Id foo] *)
-   property = (
+  (* foo.bar[1.0] -> Some ("foo", [Prop "1"; Prop "bar"]) *)
+  property = Member.PropertyExpression (_, Ast.Expression.Literal {
+    Ast.Literal.value = Ast.Literal.Number f; _; });
+  _; } -> (
+  (* The `raw` string is no good as a refinement key because of equivalence
+     classes like "1" = 1 = 0.9... = 1.0 = .... We just leave 0.9... to OCaml
+     randomness because the ES spec leaves its handling ambiguous. *)
+  let name =
```
Comment (goodmind) [inline src/typing/refinement.ml]: Should probably use `Dtoa.ecma_string_of_float`

**Comment by goodmind** [comment]
Hey @mroch. Maybe you will find this useful too for tuples

**Comment by ilya-bobyr** [comment]
Rebased this change on top of master in https://github.com/facebook/flow/pull/7974

**Comment by STRML** [comment]
It's a shame to see great PRs like this get left unreviewed and unmerged - this is a hugely useful feature.

**Comment by ilya-bobyr** [comment]
> It's a shame to see great PRs like this get left unreviewed and unmerged - this is a hugely useful feature.

We got the "length" part in, in this PR: https://github.com/facebook/flow/pull/7974

I have rebased the original change to work in compile codebase as of about a year ago.
But as @goodmind pointed out in the discussion for PR 7974, a  technical discussion might need to happen before a final implementation would go through a review.

**Comment by STRML** [comment]
Thanks for pointing that out. Length refinement is rather limited but workable. Truthiness checks and less than / greater than does not trigger the refinement. 

 See [this example](https://flow.org/try/#0FAMwrgdgxgLglgewgAhAhAKAhgLmQbQF1kAfZAQQCdKsBPAHgjAFsAjAU0oD4BKZAb2DJkcEMmwA6ADbsIAcxgALZAEIAvGuQAGPpXYwwlFFnxbCAbiEixkmfKW79h46YtXR4rNNkLlXbY4GRsgmZpYAvkA).

**Pull request closed without merging**